### DISCUSSION TITLE
[Chapter 08] MSA환경에서의 트랜잭션 동기화

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cowaine/spring-boot-for-msa#122] [Chapter 08. Data Persistence] [answered]
책에 MSA환경에서의 데이터를 처리하는 방법에
대해 나와있습니다.

Q. MSA환경에서 메세지 큐 또는 RestTemplate등으로 다른 서버로 연결된 작업들에
대해서 트랜잭션을 처리한 경험이 있다면 알려주시면 감사하겠습니다.

혹시 경험이 없다면 해당 상황에서는 어떤 방식으로 처리해야 좋을지 의견을 나눠주시면 감사하겠습니다.

**Comment by Irisation23** [comment]
Chapter 09가 RestTemplate, WebClient를 담고있기 때문에 해당 장에서 다뤄질 내용 같습니다.

**Comment by ahngilwoong** [comment, marked as answer]
선배한테 물어본 자문자답입니다.

제가 ITSM 하고있는 서비스는 컨트롤러를 통한 API 호출이 아닌 카프카를 통한 이벤트 기반 방식으로 처리를 합니다.
예를 들어 쇳물이 나오는 고로에서 쇳물이 전부 나왔다! 를 카프카 이벤트로 백엔드 서버에 넘겨줍니다. 그 이벤트가 발생하면 이제 여러 서비스에서 그 이벤트를 받는 클래스에 Transactional 어노테이션을 달고 처리합니다. 중앙 집중 처리 방식입니다.
각자의 로직을 처리하는데 이때 
try{
}안에 마이크로 서비스 함수들을 호출하고
catch에서 익셉션이 났을때 원래대로 돌이켜주는 형식이었습니다.
그런데 유지보수를 하다보면 저 try안에 서비스들이 더 추가되는데 추가된 서비스들에 대해서는 롤백이 되지 않더군요..
무언가 가라로 해놓은 느낌이 있습니다.

이게 아래 지식이 그대로 적용되어있던거였는데
좀 더 공부가 필요해 보입니다.
[보상 트랜잭션 사가패턴?](https://insanelysimple.tistory.com/368)
